vent/sqldb/sqldb.go

```
@@ -39,132 +41,198 @@ const (
-	containers := make([]string, cols+nkeys)
+// UpsertRows inserts or updates a multiple rows in PGSQL and writes _bosbarmot_log
+func (db *SQLDB) UpsertRows(tables []Table) error {
+	tx, err := db.DB.Begin()
 
-	// For Each column in Table
-	p := 0
-	cnk := 0
-	for colName, tableColumn := range table.Columns {
+	if err != nil {
+		db.log.Error("msg", "Error beginning transaction", "err", err)
+		return err
+	}
 
-		//INSERT INTO TABLE (*columns).........
-		if columns != "" {
-			columns += ", "
-			insValues += ", "
+	var table Table
+	var registers int
+
+	for registers, table = range tables {
+		query, pointers := getUpsertQuery(db.schema, table)
+
+		db.log.Debug("msg", "Upsert row", "query", clean(query), "value", pointers)
+		_, err = tx.Exec(query, pointers...)
+		if err != nil {
+			break
 		}
-		safeCol := safe(colName)
-		columns += safeCol
-		insValues += "$" + strconv.Itoa(p+1)
+	}
 
-		pointers[p] = &containers[p]
-		containers[p] = tableColumn.Value
+	if err == nil {
+		//save log
+		query := fmt.Sprintf("INSERT INTO %s._bosmarmot_log (registers, data) VALUES ($1, $2) RETURNING id", db.schema)
+		db.log.Debug("msg", "Log event", "query", clean(query), "value", registers)
+		_, err = tx.Exec(query, registers, "")
+		if err != nil {
+			db.log.Error("msg", "Error inserting to _bosmamromt_log", "err", err)
+			return err
+		}
 
-		//INSERT........... ON CONFLICT......DO UPDATE (*updValues)
-		if !tableColumn.Primary {
-			if updValues != "" {
-				updValues += ", "
+		/*
+			//get row log id
+			id, err := res.LastInsertId()
+			if err != nil {
+				db.log.Error("msg", "Error getting _bosmamromt_log id", "err", err)
+				return 0, err
 			}
-			u := cnk + cols
+		*/
 
-			updValues += safeCol + " = $" + strconv.Itoa(u+1)
-			pointers[u] = &containers[u]
-			containers[u] = tableColumn.Value
-			cnk++
+		//commit tx
+		err = tx.Commit()
+		if err != nil {
+			db.log.Error("msg", "Error committing transaction", "err", err)
```

You can probably just return the error here without logging anything, because the caller will probably log it too. The same comment applies for all the error logs in this package.